src/types/integer.rs

```diff
@@ -200,13 +200,10 @@ impl_uint_from_small_unsigned_int_types!(u8, u16, u32, u64, usize);
 macro_rules! impl_uint_try_from_small_signed_int_types {
     ($($t:ty),*) => ($(
         impl TryFrom<$t> for UInt {
-            // If conversion fails, we discard the error. `TryFromIntError` provides no
-            // additional context and cannot be constructed outside of `std`. We'll return
-            // the unit type, `()`.
-            type Error = ();
+            type Error = IonError;
             fn try_from(value: $t) -> Result<Self, Self::Error> {
                 if value < 0 {
-                    return Err(());
+                    return IonResult::illegal_operation("cannot convert a negative number to a UInt");
```

Comment:
🗺️ Previously, many conversions that would fail used `()` as their error type because the single mode of failure was obvious. (For example: trying to convert a negative `Int` to a `UInt`.) However, making these into `IonError` means they can be bubbled up with `?` alongside all other `IonResult`-producing code.

`illegal_operation` isn't ideal; I'm open to creating an `InvalidConversion` error variant, but I don't expect users to will inspect conversion errors very deeply.